hazelcast-sql/src/main/java/com/hazelcast/jet/sql/impl/parse/QueryConverter.java

```diff
@@ -156,32 +161,28 @@ private HazelcastSqlToRelConverter createSqlToRelConverter() {
      *  It is used instead of "expand" flag due to bugs in Calcite (see {@link #EXPAND}).
      * </li>
      * <li>
-     *  Transformation of distinct UNION to UNION ALL, merging the neighboring UNION relations.
+     *  Check, if the relation uses cyclic user types, and if they are allowed - skip this step.
      * </li>
      * <li>
-     *  Check, if the relation uses cyclic user types, and if they are allowed - skip this step.
+     *  Extract unsupported source expressions from an UPDATE stmt into a {@link Calc}.
      * </li>
-     *
      * </ul>
      *
+     * @implNote we're using Calcite's {@link HepProgramBuilder} to GC the rules we don't need earlier.
      * @param rel                       Initial relation.
      * @param cyclicUserTypesAreAllowed Whether cyclic user types are allowed.
      * @return Resulting relation.
      */
     private static RelNode performUnconditionalRewrites(RelNode rel, boolean cyclicUserTypesAreAllowed) {
+        // TODO: rework it at the same wat as for HEP_CALC_REWRITER_PROGRAM,
+        //  when limitation for cyclic UDTs will be removed.
         HepProgramBuilder hepProgramBuilder = new HepProgramBuilder();
 
         // Correlated subqueries elimination rules
-        hepProgramBuilder.addRuleInstance(CoreRules.FILTER_SUB_QUERY_TO_CORRELATE);
-        hepProgramBuilder.addRuleInstance(CoreRules.PROJECT_SUB_QUERY_TO_CORRELATE);
-        hepProgramBuilder.addRuleInstance(CoreRules.JOIN_SUB_QUERY_TO_CORRELATE);
-
-        // Union optimization rules
-        hepProgramBuilder.addRuleInstance(CoreRules.UNION_MERGE);
-        hepProgramBuilder.addRuleInstance(CoreRules.UNION_TO_DISTINCT);
+        hepProgramBuilder.addRuleInstance(CoreRules.FILTER_SUB_QUERY_TO_CORRELATE)
```

Comment:
this program also can be static, it does not depend on anything that changes at runtime, it only needs to know value of `cyclicUserTypesAreAllowed`, but can be created once (on first query) and then reused.

Reply:
I also thought about it, and yeah, seems it is safe to initialize it once at runtime.

Reply:
https://github.com/hazelcast/hazelcast/pull/26032/commits/a8179a6e5e9e533507a41e020d5f7bf0a5d85b47

Reply:
it causes problems in tests, the general issue is that it will cause problem when you run multiple HZ instances in the same JVM with different `cyclicUserTypesAreAllowed` setting. We had somehow similar problem recently with calcite caching something in a static map.

Reply:
@Patras3 your change will fix the problem with tests, but we support embedded HZ instances. Nothing prevents user to start many embedded instances in the same JVM with different config. Or is such usage unsupported?

Reply:
Will it work in the same way then ? I thought this problem only applies to tests and some 'magic' in surefire. 🤔